pkg/catalog/endpoint.go

```diff
@@ -38,17 +38,21 @@ func (mc *MeshCatalog) GetResolvableServiceEndpoints(svc service.MeshService) ([
 	return endpoints, nil
 }
 
-// ListAllowedEndpointsForService returns only those endpoints for a service that belong to the allowed outbound service accounts
+// ListEndpointsForServiceIdentity returns only those endpoints for a allowed outbound service accounts
 // for the given downstream identity
 // Note: ServiceIdentity must be in the format "name.namespace" [https://github.com/openservicemesh/osm/issues/3188]
-func (mc *MeshCatalog) ListAllowedEndpointsForService(downstreamIdentity identity.ServiceIdentity, upstreamSvc service.MeshService) ([]endpoint.Endpoint, error) {
+func (mc *MeshCatalog) ListEndpointsForServiceIdentity(downstreamIdentity identity.ServiceIdentity, upstreamSvc service.MeshService) ([]endpoint.Endpoint, error) {
 	outboundEndpoints, err := mc.listEndpointsForService(upstreamSvc)
 	if err != nil {
 		log.Error().Err(err).Msgf("Error looking up endpoints for upstream service %s", upstreamSvc)
 		return nil, err
 	}
+	outboundEndpointsIPSet := make(map[string]endpoint.Endpoint)
+	for _, ep := range outboundEndpoints {
+		outboundEndpointsIPSet[ep.IP.String()] = ep
```

Comment:
Isn't this logic flawed, where an endpoint's port is disregarded? Multiple endpoints with the same IP but different ports constitute multiple endpoints, but with this logic they all map to the same key.

Reply:
The current logic is also only checking the IP address: https://github.com/openservicemesh/osm/issues/3190#L63

But I agree that port should be considered during the allowing process.

Reply:
It seems like the current logic is matching on pod IP only to filer out the endpoints corresponding to the pod. But this logic is disregarding endpoints that have the same port completely.

Reply:
I have another question, why do we only compare IP addresses in the first place?

Assume 3 services, down-1, up-1, up-2. 

* down-1 ---denied--> up-1, which has endpoint 1.2.3.4:8000
* down-1 ---allowed-----> up-2, which has endpoint 1.2.3.4:9000

Thus, suppose we invoke this method `ListAllowedEndpointsForService(down-1, up-1)`, we expect it to return an empty list. In the current implementation [here](https://github.com/openservicemesh/osm/blob/main/pkg/catalog/endpoint.go#L64-L66), in the 3-level nested loop:

* `destSvcAccounts` => [up-2]
* `outboundEndpoints` => 1.2.3.4:8000 (endpoint of up-1)
* `podEndpoints` => [1.2.3.4:9000] (endpoints of up-2)

Because their IP matches, the `allowedEndpoints` => [1.2.3.4:8000]. This makes up-1 allowed. So in this example, even the current implementation is wrong. I think the current implementation has an assumption that different services will not have pods with the same IP, thus we only compared IP addresses.

Reply:
Your understanding is correct. Why not simply use a mapset for storing the endpoints and doing `podEndpoints.Intersect(allowedEndpoints)`?

Reply:
Turns out we cannot use port number as key in the hash map to match endpoints. `ListEndpointsForIdentity()` returns endpoints without port, that causes matching fail in some cases. It's detected by our CI this time.

https://github.com/openservicemesh/osm/blob/0f792a579b05de5221d81e88399a4834bd0e7c55/pkg/providers/kube/client.go#L71-L98